app-antivirus/clamav: add 0.105.1

Kangie opened the pull request:
Passes tests on `py3.11` for `~amd64`.

Had a request in IRC so I quickly bumped and tested the ebuild. It seems sane to drop the `103.*` ebuilds but please feel free to drop the commit if there's a reason to keep them around (like those local USE flags).

Edit: `freshclam` gets updates and `clamscan` can detect eicar test files.

thesamesam commented:
103.x should be kept as LTS versions.

thesamesam commented:
This doesn't seem to cater for one of the major changes with 105 - Rustification

Kangie commented:
Nothing rust in the release notes, just the llvm stuff. I'll drop 103.* commit and take a deeper look after some sleep, just a (somewhat) quick and dirty bump based on a specific request in `#gentoo`!

thesamesam commented:
https://blog.clamav.net/2022/03/clamav-01050-release-candidate-now.html?m=1

thesamesam commented on app-antivirus/clamav/clamav-0.105.1.ebuild:
```diff
@@ -0,0 +1,218 @@
+	rar? ( app-arch/unrar )
+	test? ( dev-python/pytest )"
+# TODO: there is no way to use this with the new build system instead of the bundled one
+#	dev-libs/tomsfastmath
```
105 lets us sort this: "ENABLE_EXTERNAL_TOMSFASTMATH"

Kangie commented:
Yeah, I found it eventually. I generally look for big ticket items like that in the actual version release not the RC!

Good news, it looks looks like CMake finds `rustc` no worries and the rust dependencies are all vendored. I actually didn't even notice that rust was an additional dep on an up-to-date system. 

Also thanks for the colour stuff on the wiki; build logs are *so* much more readable.

Not sure if tomsfastmath was supposed to be a `BDEPEND` or `CDEPEND` based on that comment (one of the reasons that I put whitespace between those envvars); I'm not going to second guess where I put it at 1 AM, let me know if I'm wrong

Kangie commented:
As discussed on IRC, the JIT flag fails with (unclear from upstream) but llvm >13; and their CMake implementation doesn't seem to be respecting their documented override flags. WIP!

thesamesam commented on app-antivirus/clamav/clamav-0.105.1.ebuild:
```diff
@@ -0,0 +1,233 @@
+		-DCLAMAV_GROUP="clamav"
+		-DLLVM_ROOT_DIR="$(get_llvm_prefix -d ${LLVM_MAX_SLOT})"
+		-DLLVM_FIND_VERSION="$(best_version sys-devel/llvm:${LLVM_MAX_SLOT})"
+		-DBYTECODE_RUNTIME=$(usex jit llvm interpreter)
```
You dropped the comments but I think they're worth having -- both for users who get curious as well as maintainers.

Kangie replied:
Couldn't get your patch to apply cleanly; consider this a WIP in case of SIGCAT (which I've prevented four times already since I sat back down...)

LordVan commented:
I thought I commented already last night, but either I put it in the wrong place, or I was too tired and failed to send. Anyway. to make it short: thanks for the testing this rust build stuff ,.. I had not had the time to properly test it (also because I know too little about rust and gentoo's implementation )

here's the bugzie bug linked btw https://bugs.gentoo.org/842852

Kangie commented:
OK, I've patched `CMakelists.txt` to (I hope...) appropriately search for the correct versions of LLVM and added some sanity checks for 'this version is too high to work'. I'd appreciate some feedback - JIT is not currently building with 13.0.1 (the linker is not happy) but it passes the LLVM compile stage so I'm going to say that I've cracked that portion.

I'll quickly add a package.use.mask so that we can merge this if we're happy but if someone feels like looking over the build log (etc) it might enlighten me as to what *else* needs to be fixed...

Also my CMake-fu is not strong; if I've gone about it backwards I'll accept any feedback on that patch before upstreaming it. :)

Kangie commented:
I've dug a little further into this; as far as I can tell something *else* is broken upstream.

LLVM detection works and (with the amended patch) is now reliable; I'll see about upstreaming that later on today.

What I get from trying a few builds is that the LLVM libraries are properly detected, stored in a variable, and passed to the linker... which has no idea where these files are and promptly fails.

The CMake docs say that paths coming out of `find_package()` are absolute and can be passed directly into `target_link_libraries()`. `LLVM_INCLUDE_DIRS` has been added to the to the include path for the bytecode runtime; but that's evidently not working either.

Tracing that further, `LLVM_INCLUDE_DIRS` is either set by `FindLLVM.cmake` (included but not used by modern LLVM) or exported by `LLVMConfig.cmake` (Installed with LLVM 13/14 at least) - which is what my system uses while configuring. `LLVM_INCLUDE_DIRS` is also set to the right include path... so why is the linker upset?

If it wasn't for Cisco-Talos/clamav/issues/574 I'd be prepared to write it off as "just plain broken"

My LLVM experience is effectively 0, and my CMake experience isn't much greater so I'd appreciate any advice.

LordVan commented:
> If it wasn't for [Cisco-Talos/clamav/issues/574](https://github.com/Cisco-Talos/clamav/issues/574) I'd be prepared to write it off as "just plain broken"

Yeah ClamAV & LLVM have given me quite some headaches in the past before .. at some point there were even versions where it was just disabled because there were no LLVM versions in gentoo anymore that would work with it.. I'll try to find some time to look into it later, but LLVM is not something I know too much about either tbh (apart from the clamav headaches with it)

Kangie commented:
Thought I left this comment already!

Moved CMake LLVM override options behind `test` use flag and removed the test override for `valgrind` in the configure stage - I can't see that it's currently doing anything. Re-enable it if anyone flags an issue on an arch other than `~amd64`. Also moved the `python_has_version` check for test deps behind the `test` use flag; I'm sure that would have caused issues down the line.

Patch is in-line with what I've submitted upstream and I'm happy that the ebuild machinery is here for `jit` support in future. Still no luck with getting LLVM library linking to succeed, but I'm happy to amend that down the line if I get a response.

I'm happy enough with this ebuild to say that we go ahead and merge at this point; `jit` will need to be masked on other arches as this package is keyworded.

thesamesam commented:
> Thought I left this comment already!
> 
> Moved CMake LLVM override options behind `test` use flag 

Why? It's all dependent on JIT.

(Oh, I see it is USE=jit, and it's a typo)

> and removed the test override for `valgrind` in the configure stage - I can't see that it's currently doing anything. 

It is - `find_package(Valgrind)` is still in the codebase. Besides, we don't generally find Valgrind tests valuable to run in ebuilds.

> 
> Patch is in-line with what I've submitted upstream and I'm happy that the ebuild machinery is here for `jit` support in future. Still no luck with getting LLVM library linking to succeed, but I'm happy to amend that down the line if I get a response.
> 
> I'm happy enough with this ebuild to say that we go ahead and merge at this point; `jit` will need to be masked on other arches as this package is keyworded.

Yeah, me too.

thesamesam commented on app-antivirus/clamav/clamav-0.105.1.ebuild:
```diff
@@ -0,0 +1,236 @@
+	selinux? ( sec-policy/selinux-clamav )"
+
+python_check_deps() {
+	use test && python_has_version -b "dev-python/pytest[${PYTHON_USEDEP}]"
```
You can drop the use test && here now.

Kangie replied:
> It is - `find_package(Valgrind)` is still in the codebase. Besides, we don't generally find Valgrind tests valuable to run in ebuilds.

Sorry, should have been more clear; It's passing tests in the sandbox and although the comment indicates: 

```bash
		# Used to enable some more tests but doesn't behave well in
		# sandbox necessarily(?) + needs certain debug symbols present
		# in e.g. glibc.
```

It doesn't actually seem to enable any additional tests.

I'm happy to put it back in behind the `use test` conditional, I just can't see that it's actually having any measurable impact; I can't think of a better way to validate that than seeing if it throws errors as people test and keyword on systems that aren't mine.

Kangie commented:
> It is - `find_package(Valgrind)` is still in the codebase. Besides, we don't generally find Valgrind tests valuable to run in ebuilds.

Sorry, should have been more clear; It's passing tests in the sandbox and although the comment indicates: 

```bash
		# Used to enable some more tests but doesn't behave well in
		# sandbox necessarily(?) + needs certain debug symbols present
		# in e.g. glibc.
```

It doesn't actually seem to enable any additional tests.

I'm happy to put it back in behind the `use test` conditional, I just can't see that it's actually having any measurable impact; I can't think of a better way to validate that than seeing if it throws errors as people test and keyword on systems that aren't mine.

LordVan commented:
Actually one question @Kangie why did you drop all ~arch except amd64? the only ones not satisfying the dependencies are hppa and ia64 from what I can see - or did I miss something?

Kangie commented:
@LordVan I currently can only test on `~amd64` (though my other arch hardware collection is coming along...).

With the changes that I've made here I figured that we'd be better off getting the package re-keyworded for testing.

Besides, iirc, we're not supposed to keyword a version bump for an arch that it hasn't been tested on. :)

thesamesam commented:
> @LordVan I currently can only test on `~amd64` (though my other arch hardware collection is coming along...).
> 
> With the changes that I've made here I figured that we'd be better off getting the package re-keyworded for testing.
> 
> Besides, iirc, we're not supposed to keyword a version bump for an arch that it hasn't been tested on. :)

No, see https://devmanual.gentoo.org/keywording/index.html#keywording-on-upgrades.

thesamesam commented:
> > It is - `find_package(Valgrind)` is still in the codebase. Besides, we don't generally find Valgrind tests valuable to run in ebuilds.
> 
> Sorry, should have been more clear; It's passing tests in the sandbox and although the comment indicates:
> 
> ```shell
> 		# Used to enable some more tests but doesn't behave well in
> 		# sandbox necessarily(?) + needs certain debug symbols present
> 		# in e.g. glibc.
> ```
> 
> It doesn't actually seem to enable any additional tests.
> 
> I'm happy to put it back in behind the `use test` conditional, I just can't see that it's actually having any measurable impact; I can't think of a better way to validate that than seeing if it throws errors as people test and keyword on systems that aren't mine.

DId you look at the bug it was added for? 

With Valgrind installed:
```
[...]
-- Found Python3: /usr/sbin/python3.11 (found version "3.11.0") found components: Interpreter
-- Valgrind found: /usr/sbin/valgrind, 3.19.0
-- Found Threads: TRUE
[...]

[...]

Test project /var/tmp/portage/app-antivirus/clamav-0.105.1/work/clamav-0.105.1_build
      Start  1: libclamav
      Start  2: libclamav_valgrind
      Start  3: libclamav_rust
      Start  4: clamscan
      Start  5: clamscan_valgrind
      Start  6: clamd
      Start  7: clamd_valgrind
      Start  8: freshclam
      Start  9: freshclam_valgrind
      Start 10: sigtool
      Start 11: sigtool_valgrind
 1/11 Test #10: sigtool ..........................   Passed    1.04 sec
[...]
```

It's clearly running additional tests when Valgrind is around. We don't want that -- valgrind tests are flaky under sandbox, and it's non-portable, per the bug which led to it being added. Please undo that part of the changes.

Kangie commented:
Undid that also:

-  added `USE=experimental` to toggle experimental features, because "why not?" - this is Gentoo. If someone chooses to shoot themselves in the foot they have every right to do so using portage.
- OpenRC runlevel postinst messages now hidden behind `if ! systemd_is_booted`

If there's any other little stuff like that I'm happy to implement.

thesamesam commented on profiles/features/wd40/package.mask:
```diff
@@ -4,6 +4,7 @@
 # Various packages requiring Rust.
 app-admin/ansible
 app-admin/ansible-core
+app-antivirus/clamav
```
mask >= 0.105.0 instead, although indeed it's going to be a problem when LTS is discontinued :(

thesamesam commented on app-antivirus/clamav/clamav-0.105.1.ebuild:
```diff
@@ -0,0 +1,246 @@
+		# Used to enable some more tests but doesn't behave well in
+		# sandbox necessarily(?) + needs certain debug symbols present
+		# in e.g. glibc.
+		mycmakeargs+=(-DCMAKE_DISABLE_FIND_PACKAGE_Valgrind=ON)
```
Nit:
```suggestion
		mycmakeargs+=(  -DCMAKE_DISABLE_FIND_PACKAGE_Valgrind=ON )
```

LordVan commented:
Tested again and looks good to me .. 
I would merge this probably tomorrow unless you found any other issues @thesamesam

thesamesam commented:
All good from my POV, thanks!
